File: Algorithms/min_max/min_max.py
import matplotlib.pyplot as plt
import numpy as np
from typing import List
import math
# Plot a square given the center and the side length / 2

class Rectangle:
	def __init__(self, left_bot, right_top):
		self.xmin, self.ymin = left_bot
		self.xmax, self.ymax = right_top
	
	def print_coords(self):
		print(f"Left bottom is {self.xmin, self.ymin}")
		print(f"Right top is {self.xmax, self.ymax}")

def make_rect(r: Rectangle) -> tuple[List[float], List[float]]:
	"""
	Returns the (x, y) co-ordinates of the square
	"""
	x_min, y_min = r.xmin, r.ymin
	x_max, y_max = r.xmax, r.ymax
	num_pts = 1000

	x_points = []
	y_points = []

	# Bottom Line 
	x_points = x_points + list(np.linspace(x_min, x_max, num_pts))
	y_points = y_points + list(np.linspace(y_min, y_min, num_pts))

	# Right Line 
	x_points = x_points + list(np.linspace(x_max, x_max, num_pts))
	y_points = y_points + list(np.linspace(y_min, y_max, num_pts))

	# Top line
	x_points = x_points + list(np.linspace(x_max, x_min, num_pts))
	y_points = y_points + list(np.linspace(y_max, y_max, num_pts))

	# Left Line
	x_points = x_points + list(np.linspace(x_min, x_min, num_pts))
	y_points = y_points + list(np.linspace(y_max, y_min, num_pts))

	return (x_points, y_points)

def euclidean_dist(pos1: tuple[float, float], pos2: tuple[float,float]) -> float:
	"""
	Returns the Euclidean Distance between position1 and position2
	(x1 - x2)**2 + (y1 - y2)**2
	"""
	x1, y1 = pos1
	x2, y2 = pos2
	return  math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)

def get_distances(centers: List[tuple[float, float]]) -> List[float]:

	random_pt = (0.5, 0.5)
	distances = []
	for center in centers:
		distances.append(euclidean_dist(center, random_pt))
	return random_pt, distances

def get_intersection(r1: Rectangle, r2: Rectangle) -> Rectangle:
	"""
	Gets the Intersection Rectangle's left bottom and right top co - ordinates
	"""
	x1, y1 = max(r1.xmin, r2.xmin), max(r1.ymin, r2.ymin)
	x2, y2 = min(r1.xmax, r2.xmax), min(r1.ymax, r2.ymax)

	assert(x1 <= x2)
	assert(y1 <= y2)

	return Rectangle((x1, y1), (x2, y2))

def plot_rect(r: Rectangle) -> None:
	co_ords = make_rect(r)
	plt.scatter(co_ords[0], co_ords[1], s = 3)

if __name__ == "__main__":
	centers = [(0, 0), (1, 0), (0.5, 1)]
	rectangles = []
	random_pt, distances = get_distances(centers)

	for center, distance in zip(centers, distances):
		x_min, x_max = center[0] - distance, center[0] + distance
		y_min, y_max = center[1] - distance, center[1] + distance
		r = Rectangle((x_min, y_min), (x_max, y_max))
		rectangles.append(r)
		plot_rect(r)

	r1 = get_intersection(rectangles[0], rectangles[1])
	r_final = get_intersection(rectangles[2], r1)
	plot_rect(r_final)
	plt.scatter(random_pt[0], random_pt[1], s = 12)
	plt.show()

File: Visualisations/static/min_max.js
class Rectangle {

	constructor(xmin, ymin, xmax, ymax){
		this.xmin = xmin
		this.ymin = ymin
		this.xmax = xmax
		this.ymax = ymax
	}
	
	print_coords(){
		console.log("Left bottom is: ", this.xmin, ", ", this.ymin);
		console.log("Right top is: ", this.xmax, ", ", this.ymax);
	}

}

function euclidean_distance(x1, y1, x2, y2){

	return Math.sqrt(Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2))

}

function get_intersection(r1, r2){

	x1 = Math.max(r1.xmin, r2.xmin);
	y1 = Math.max(r1.ymin, r2.ymin);
	x2 = Math.min(r1.xmax, r2.xmax);
	y2 = Math.min(r1.ymax, r2.ymax);

	console.assert(x1 <= x2);
	console.assert(y1 <= y2);

	let rect = new Rectangle(x1, y1, x2, y2);
	return rect;

}

function min_max_calc(myData){

	let centers = [[myData.x1, myData.y1], [myData.x2, myData.y2], [myData.x3, myData.y3]];
    let distances = [myData.d1, myData.d2, myData.d3];
	
	let rectangles = []

	let len1 = centers.length;
	let len2 = distances.length;
	console.assert(len1 == len2);
	for(let i = 0; i < len1; i++){
		center = centers[i];
		distance = distances[i];
		let r = new Rectangle(center[0] - distance, center[1] - distance, center[0] + distance, center[1] + distance)
		rectangles.push(r)

	}

	r1 = get_intersection(rectangles[0], rectangles[1]);
	r_final = get_intersection(rectangles[2], r1);
	r_final.print_coords();

}

function handleData(){

	let x = 0.5
	let y = 0.5
	console.log("My dummy point is ", x, ", ", y);

	let myData = {
		x1: 0,
		y1: 0,
		x2: 1,
		y2: 0,
		x3: 0.5,
		y3: 1
	}
	myData.d1 = euclidean_distance(myData.x1, myData.y1, x, y)
	myData.d2 = euclidean_distance(myData.x2, myData.y2, x, y)
	myData.d3 = euclidean_distance(myData.x3, myData.y3, x, y)
	console.log(myData)
	result_rect = min_max_calc(myData)

}

console.log("Javascript was called");
handleData()
